etc/profile-m-z/nextcloud-desktop.profile

```diff
@@ -0,0 +1,48 @@
+# Firejail profile for nextcloud-desktop
+# Description: Nextcloud desktop client
+# This file is overwritten after every install/update
+# Persistent local customizations
+include nextcloud-desktop.local
+# Persistent global definitions
+include globals.local
+
+noblacklist ${DOCUMENTS}
+noblacklist ${HOME}/Nextcloud
+noblacklist ${HOME}/.config/Nextcloud
+noblacklist ${HOME}/.local/share/Nextcloud
+
+include disable-common.inc
+include disable-devel.inc
+include disable-interpreters.inc
+include disable-passwdmgr.inc
+include disable-programs.inc
+
+mkdir ${HOME}/Nextcloud
+mkdir ${HOME}/.config/Nextcloud
+mkdir ${HOME}/.local/share/Nextcloud
+whitelist ${DOCUMENTS}
+whitelist ${HOME}/Nextcloud
+whitelist ${HOME}/.config/Nextcloud
+whitelist ${HOME}/.local/share/Nextcloud
+include whitelist-common.inc
```

Comment:
I'm not sure if it should be an whitelisting profile, but I also don't know what features nextcloud-desktop has.

Reply:
```
ls ~/.config/Nextcloud/
cookies0.db  cookies1.db  logs  nextcloud.cfg  sync-exclude.lst
```

```
ls ~/.local/share/Nextcloud/
Nextcloud_sync.log
```

Reply:
> Client fails to login without the following 2 lines:

When you remove every `whitelist ${HOME}/...` and `include whitelist-common.inc`?

Reply:
> Client fails to login without the following 2 lines:

When you remove every `whitelist ${HOME}/...` and `include whitelist-common.inc`?

Reply:
No. No login in this way
```
#whitelist ${HOME}/Nextcloud
#whitelist ${HOME}/.config/Nextcloud
#whitelist ${HOME}/.local/share/Nextcloud
include whitelist-common.inc
```

It seems we really need those lines

Reply:
Because whitelisting is still enabled, you need to comment `include whitelist-common.inc` too to have a blacklisting profile.

Reply:
>you need to comment include whitelist-common.inc too 

sorry, now I did. Unfortunately same result. No login or no sync.

---
`whitelist ${HOME}/Nextcloud` is the defaults sync folder (user can change and use what he wants, of course)
`whitelist ${DOCUMENTS}` I would let as a commented option. Can be enabled in case someone wants to sync another directory in Documents
`whitelist ${HOME}/.config/Nextcloud` and `whitelist ${HOME}/.local/share/Nextcloud` are used for Nextcloud login and settings

I really don't know how to go around those whitelists

Maybe we could propose a more safe location for the Nextcloud sync directory, but what to do with config and local? We really need those

Reply:
We can skip the mkdir but (still) not whitelist. Tell me if you think could be better.

```
#mkdir ${HOME}/Nextcloud
#mkdir ${HOME}/.config/Nextcloud
#mkdir ${HOME}/.local/share/Nextcloud
# Uncomment or put in your nextcloud-desktop.local to allow sync with Document directory.
#whitelist ${DOCUMENTS}
whitelist ${HOME}/Nextcloud
whitelist ${HOME}/.config/Nextcloud
whitelist ${HOME}/.local/share/Nextcloud
```

Reply:
Are you sure that there is no "hidden" `whitelist`? (check with `firejail --ignore=private-bin --profile=nextcloud-desktop ls -lA $HOME`).

> We can skip the mkdir

Nope, because you can only `whitelist` file which exist, this will break if `${HOME}/.config/Nextcloud`/`${HOME}/.local/share/Nextcloud` are not present (e.g. first start).

Reply:
```
$ firejail --ignore=private-bin --profile=nextcloud-desktop ls -1A $HOMEReading profile /home/user/.config/firejail/nextcloud-desktop.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-passwdmgr.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/disable-shell.inc
Reading profile /etc/firejail/disable-xdg.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-runuser-common.inc
Reading profile /etc/firejail/whitelist-usr-share-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Warning: networking feature is disabled in Firejail configuration file
Parent pid 28736, child pid 28739
Warning: An abstract unix socket for session D-BUS might still be available. Use --net or remove unix from --protocol set.
Blacklist violations are logged to syslog
Child process initialized in 231.94 ms
.bashrc
.cache
.config
.local
Nextcloud
.themes

Parent is shutting down, bye...
```

Reply:
> Reading profile /etc/firejail/whitelist-common.inc

Can you try with every whitelist commented or maybe even with `--ignore='whitelist ${HOME}'`.

Reply:
Found!
Is this way good enough?

```
# Uncomment or put in your nextcloud-desktop.local to allow sync with Document directory.
#noblacklist ${DOCUMENTS}
noblacklist ${HOME}/Nextcloud
noblacklist ${HOME}/.config/Nextcloud
noblacklist ${HOME}/.local/share/Nextcloud

include disable-common.inc
include disable-devel.inc
include disable-exec.inc
include disable-interpreters.inc
include disable-passwdmgr.inc
include disable-programs.inc
include disable-shell.inc
include disable-xdg.inc

mkdir ${HOME}/Nextcloud
mkdir ${HOME}/.config/Nextcloud
mkdir ${HOME}/.local/share/Nextcloud
# Uncomment or put in your nextcloud-desktop.local to allow sync with Document directory.
#whitelist ${DOCUMENTS}
#whitelist ${HOME}/Nextcloud
#whitelist ${HOME}/.config/Nextcloud
#whitelist ${HOME}/.local/share/Nextcloud
#include whitelist-common.inc
#include whitelist-runuser-common.inc
#include whitelist-usr-share-common.inc
#include whitelist-var-common.inc
```

Reply:
Maybe we could go with
```
# Uncomment (or put in your nextcloud-desktop.local) if you want to sync from these directories.
#blacklist ${DOCUMENTS}
#blacklist ${MUSIC}
#blacklist ${PICTURES}
#blacklist ${VIDEOS}
noblacklist ${HOME}/Nextcloud
noblacklist ${HOME}/.config/Nextcloud
noblacklist ${HOME}/.local/share/Nextcloud

include disable-common.inc
include disable-devel.inc
include disable-exec.inc
include disable-interpreters.inc
include disable-passwdmgr.inc
include disable-programs.inc
include disable-shell.inc
include disable-xdg.inc

# Uncomment the following (or put in your nextcloud-desktop.local) to enable extra protection.
#mkdir ${HOME}/Nextcloud
#mkdir ${HOME}/.config/Nextcloud
#mkdir ${HOME}/.local/share/Nextcloud
#whitelist ${HOME}/Nextcloud
#whitelist ${HOME}/.config/Nextcloud
#whitelist ${HOME}/.local/share/Nextcloud
#include whitelist-common.inc
include whitelist-runuser-common.inc
include whitelist-usr-share-common.inc
include whitelist-var-common.inc
```

Reply:
**It works! **

We were on the same way in the same time :-)
I have a big delay to receive notifications
You added more nice stuff

A question: 

>Uncomment the following (or put in your nextcloud-desktop.local) to enable extra protection.

I thought it's better to have less whitelists as possible. That's the reason we were working on. Did I understood wrong?

Reply:
> #Uncomment (or put in your nextcloud-desktop.local) if you want to sync from these directories.
> #blacklist ${DOCUMENTS}
> #blacklist ${MUSIC}
> #blacklist ${PICTURES}
> #blacklist ${VIDEOS}

Shouldn't be all 
`#noblacklist ...`

Reply:
So, now, only this part is not really clear to me.

To enable extra protection we can enable those lines? I thought we were working to kick away the whitelists
And mkdirs are important if a dir is not yet there.
Maybe I really understood something different
```
# Uncomment the following (or put in your nextcloud-desktop.local) to enable extra protection.
#mkdir ${HOME}/Nextcloud
#mkdir ${HOME}/.config/Nextcloud
#mkdir ${HOME}/.local/share/Nextcloud
#whitelist ${HOME}/Nextcloud
#whitelist ${HOME}/.config/Nextcloud
#whitelist ${HOME}/.local/share/Nextcloud
#include whitelist-common.inc
include whitelist-runuser-common.inc
include whitelist-usr-share-common.inc
include whitelist-var-common.inc
```

Reply:
> I thought it's better to have less whitelists as possible. That's the reason we were working on. Did I understood wrong?

less whitelist = less permissions but no whitelist = full permissions (minus blacklisted permissions). Whitelisting is much much more safer the blacklisting. On the other hand you can only sync into/from whitelisted directories. I'm still not sure if we should even comment the disable-xdg or enable whitelisting. What's the usuall workflow? How likely is that it disturbs? If anyone else can say something too.

Reply:
I did more tests about those 2 choices
1) If I use noblacklist, mkdir and whitelist, I cannot choose what to sync from the server and all is always downloaded 
2) if I use only noblacklist (like it is now), all /home is visible and I can sync all directories I want

I think, the first we cannot use, sync is really a pain.
The second, we need a kind of restriction
This is the second
```
# Uncomment or put in your nextcloud-desktop.local to allow sync with Document directory.
#noblacklist ${DOCUMENTS}
#noblacklist ${MUSIC}
#noblacklist ${PICTURES}
#noblacklist ${VIDEOS}
noblacklist ${HOME}/Nextcloud
noblacklist ${HOME}/.config/Nextcloud
noblacklist ${HOME}/.local/share/Nextcloud

include disable-common.inc
include disable-devel.inc
include disable-exec.inc
include disable-interpreters.inc
include disable-passwdmgr.inc
include disable-programs.inc
include disable-shell.inc
include disable-xdg.inc

# Uncomment the following (or put in your nextcloud-desktop.local) to enable extra protection.
#mkdir ${HOME}/Nextcloud
#mkdir ${HOME}/.config/Nextcloud
#mkdir ${HOME}/.local/share/Nextcloud
#whitelist ${HOME}/Nextcloud
#whitelist ${HOME}/.config/Nextcloud
#whitelist ${HOME}/.local/share/Nextcloud
#include whitelist-common.inc
include whitelist-runuser-common.inc
include whitelist-usr-share-common.inc
include whitelist-var-common.inc
```